lib/layer.cpp

```diff
@@ -0,0 +1,447 @@
+    break;
+  default:
+    return tdispatch->clGetDeviceInfo(device, param_name, param_value_size,
+                                      param_value, param_value_size_ret);
+    break;
+  }
+}
+
+cl_program layer::clCreateProgramWithIL(cl_context context, const void *il,
+                                        size_t length, cl_int *errcode_ret) {
+  spirv2clc::translator translator;
+  std::string srcgen;
+  std::vector<uint32_t> binary(static_cast<const uint32_t *>(il),
+                               static_cast<const uint32_t *>(il) +
+                                   length / sizeof(uint32_t));
+  int err = translator.translate(binary, &srcgen);
+
+  if (err == 0) {
+    const char *str = srcgen.data();
+    size_t len = srcgen.length();
+    cl_program prog = tdispatch->clCreateProgramWithSource(context, 1, &str,
+                                                           &len, errcode_ret);
+    if (*errcode_ret == CL_SUCCESS) {
+      program_ils[prog] =
+          std::make_pair(std::string(static_cast<const char *>(il), length),
+                         std::string(str, len));
+    }
+    return prog;
+  } else {
+    return tdispatch->clCreateProgramWithIL(context, il, length, errcode_ret);
+  }
+}
+
+cl_int layer::clBuildProgram(cl_program program, cl_uint num_devices,
+                             const cl_device_id *device_list,
+                             const char *options,
+                             void(CL_CALLBACK *pfn_notify)(cl_program program,
+                                                           void *user_data),
+                             void *user_data) {
+  return tdispatch->clBuildProgram(program, num_devices, device_list, options,
+                                   pfn_notify, user_data);
+}
+
+cl_int layer::clGetProgramInfo(cl_program program, cl_device_info param_name,
+                               size_t param_value_size, void *param_value,
+                               size_t *param_value_size_ret) {
+  switch (param_name) {
+  case CL_PROGRAM_IL:
+    return clGetProgramInfo_CL_PROGRAM_IL(program, param_name, param_value_size,
+                                          param_value, param_value_size_ret);
+    break;
+  default:
+    return tdispatch->clGetProgramInfo(program, param_name, param_value_size,
+                                       param_value, param_value_size_ret);
+    break;
+  }
+}
+
+cl_int layer::clGetDeviceInfo_CL_DEVICE_EXTENSIONS(
+    cl_device_id device, cl_device_info param_name, size_t param_value_size,
+    void *param_value, size_t *param_value_size_ret) {
+  cl_int err_;
+  bool layer_has_nothing_to_do =
+      device_supports_spirv_out_of_the_box(device, &err_);
+  if (layer_has_nothing_to_do || err_ != CL_SUCCESS) {
+    return tdispatch->clGetDeviceInfo(device, CL_DEVICE_EXTENSIONS,
+                                      param_value_size, param_value,
+                                      param_value_size_ret);
+  } else {
```

Comment:
The else and nesting are unnecessary.

Reply:
Removed.

Reply:
Just to be clear: that comment applies to several locations in the code. I just pointed out one occurrence.